### ISSUE TITLE
slice method fails with this expr `pl.col(["a","b"]).slice(0,pl.col("a").len())` after rust-polars  0.31.0

### ISSUE THREAD

**Issue by sorhawell** [pola-rs/polars#11647]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
In [1]: %set_env POLARS_VERBOSE=1
env: POLARS_VERBOSE=1

In [2]: import polars as pl

In [3]: pl.LazyFrame({"a":[0,1,2,3,4,5], "b":[0,1,2,3,4,5]}).select(pl.col(["a"]).slice(0,pl.col("a").len())).collect()
Out[3]: 
shape: (6, 1)
┌─────┐
│ a   │
│ --- │
│ i64 │
╞═════╡
│ 0   │
│ 1   │
│ 2   │
│ 3   │
│ 4   │
│ 5   │
└─────┘

In [4]: pl.LazyFrame({"a":[0,1,2,3,4,5], "b":[0,1,2,3,4,5]}).select(pl.col(["a","b"]).slice(0,6)).collect()
Out[4]: 
shape: (6, 2)
┌─────┬─────┐
│ a   ┆ b   │
│ --- ┆ --- │
│ i64 ┆ i64 │
╞═════╪═════╡
│ 0   ┆ 0   │
│ 1   ┆ 1   │
│ 2   ┆ 2   │
│ 3   ┆ 3   │
│ 4   ┆ 4   │
│ 5   ┆ 5   │
└─────┴─────┘

#WRONG
In [5]: pl.LazyFrame({"a":[0,1,2,3,4,5], "b":[0,1,2,3,4,5]}).select(pl.col(["a","b"]).slice(0,pl.col("a").len())).collect()
Out[5]: 
shape: (1, 2)
┌─────┬─────┐
│ a   ┆ b   │
│ --- ┆ --- │
│ u32 ┆ u32 │
╞═════╪═════╡
│ 6   ┆ 6   │
└─────┴─────┘
```

### Issue description

First observed in r-polars unittest when bumping dependency rust-polars 0.31 -> 0.33.

### Expected behavior

slice should produce the same number of rows, no matter  pl.col("a") or pl.col(["a","b"]) or if length-arg is a literal or derived Expr.

### Installed versions

<details>

```
In [6]: pl.show_versions()
--------Version info---------
Polars:              0.19.7
Index type:          UInt32
Platform:            macOS-12.6.6-x86_64-i386-64bit
Python:              3.11.4 (main, Jul 25 2023, 17:08:27) [Clang 14.0.0 (clang-1400.0.29.202)]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              <not installed>
gevent:              <not installed>
matplotlib:          <not installed>
numpy:               <not installed>
openpyxl:            <not installed>
pandas:              <not installed>
pyarrow:             <not installed>
pydantic:            <not installed>
pyiceberg:           <not installed>
pyxlsb:              <not installed>
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>
```

</details>

**Comment by cmdlineluser** [comment]
If it helps debugging, the first thing I try is to disable optimization:
```python
(pl.LazyFrame({"a":[0,1,2,3,4,5], "b":[0,1,2,3,4,5]})
   .select(pl.col("a","b").slice(0,pl.col("a").len()))
   .collect(no_optimization=True)
)

# shape: (6, 2)
# ┌─────┬─────┐
# │ a   ┆ b   │
# │ --- ┆ --- │
# │ i64 ┆ i64 │
# ╞═════╪═════╡
# │ 0   ┆ 0   │
# │ 1   ┆ 1   │
# │ 2   ┆ 2   │
# │ 3   ┆ 3   │
# │ 4   ┆ 4   │
# │ 5   ┆ 5   │
# └─────┴─────┘
```
In this case, it seems to be a CSE issue:
```python
(pl.LazyFrame({"a":[0,1,2,3,4,5], "b":[0,1,2,3,4,5]})
   .select(pl.col("a","b").slice(0,pl.col("a").len()))
   .collect(comm_subexpr_elim=False)
)

# shape: (6, 2)
# ┌─────┬─────┐
# │ a   ┆ b   │
# │ --- ┆ --- │
# │ i64 ┆ i64 │
# ╞═════╪═════╡
# │ 0   ┆ 0   │
# │ 1   ┆ 1   │
# │ 2   ┆ 2   │
# │ 3   ┆ 3   │
# │ 4   ┆ 4   │
# │ 5   ┆ 5   │
# └─────┴─────┘
```

**Comment by cmdlineluser** [comment]
@reswqa  Looks to be fixed on `0.19.8` 

```python
[ins] In [3]: pl.__version__,  (pl.LazyFrame({"a":[0,1,2,3,4,5], "b":[0,1,2,3,4,5]})
         ...:    .select(pl.col(["a","b"]).slice(0,pl.col("a").len()))
         ...:    .collect()
         ...: )
Out[3]: 
('0.19.8',
 shape: (6, 2)
 ┌─────┬─────┐
 │ a   ┆ b   │
 │ --- ┆ --- │
 │ i64 ┆ i64 │
 ╞═════╪═════╡
 │ 0   ┆ 0   │
 │ 1   ┆ 1   │
 │ 2   ┆ 2   │
 │ 3   ┆ 3   │
 │ 4   ┆ 4   │
 │ 5   ┆ 5   │
 └─────┴─────┘)
```
